src/targets/maven.ts

```diff
@@ -102,48 +93,115 @@ export class MavenTarget extends BaseTarget {
-    await withTempDir(
-      async dir => {
-        const git = simpleGit(dir);
-        const username = await getAuthUsername(this.github);
-        this.githubRemote.setAuth(username, getGithubApiToken());
-        await git.clone(this.githubRemote.getRemoteStringWithAuth(), dir);
-        await git.checkout(`release/${version}`); // TODO: release name should be customized
-
-        await this.createUserGradlePropsFile();
-        const mavenReleaser = new MavenReleaser(
-          this.mavenConfig.distributionsPath,
-          this.mavenConfig.settingsPath,
-          this.mavenConfig.mavenRepoUrl,
-          this.mavenConfig.mavenRepoId,
-          this.mavenConfig.mavenCliPath,
-          this.mavenConfig.gradleCliPath
-        );
-        mavenReleaser.release();
-
-        await git.add(FILES_TO_COMMIT);
-        await git.commit(`craft(maven): Deployed ${version} to Maven Central.`);
-        if (this.shouldPush()) {
-          await git.push();
+  public async publish(_version: string, _revison: string): Promise<void> {
+    this.createUserGradlePropsFile();
+    if (isDryRun()) {
+      this.logger.info('[dry-run] Not uploading to Maven.');
+      return;
+    }
+    this.upload();
+    this.closeAndRelease();
+  }
+
+  /**
+   * Deploys to Maven Central the distribution packages.
+   * Note that after upload, this must be `closeAndRelease`.
+   */
+  public upload(): void {
+    const distributionsDirs = fs.readdirSync(
+      this.mavenConfig.distributionsPath
+    );
+    for (const distDir of distributionsDirs) {
+      const moduleName = path.parse(distDir).base;
+      const androidFile = this.getAndroidDistributionFile(distDir);
+      const targetFile = androidFile
+        ? androidFile
+        : path.join(distDir, `${moduleName}.jar`);
+      const javadocFile = path.join(distDir, `${moduleName}-javadoc.jar`);
+      const sourcesFile = path.join(distDir, `${moduleName}-sources.jar`);
+      const pomFile = path.join(distDir, 'pom-default.xml');
+
+      const command = this.getMavenUploadCmd(
+        targetFile,
+        javadocFile,
+        sourcesFile,
+        pomFile
+      );
+
+      withRetry(async () => {
+        exec(command, (error, _stdout, _stderr) => {
+          throw new Error(`Cannot upload ${distDir} to Maven:\n` + error);
+        });
+      });
+    }
+  }
+
+  /**
+   * Finishes the release flow.
+   */
+  public closeAndRelease(): void {
+    withRetry(async () => {
+      exec(
+        `./${this.mavenConfig.gradleCliPath} closeAndReleaseRepository`,
+        (error, _stdout, _stderr) => {
+          throw new Error(`Cannot close and release to Maven:\n` + error);
         }
-      },
-      true,
-      'craft-release-maven-'
+      );
+    });
+  }
+
+  /**
+   * Returns the path to the first Android distribution file, if any.
+   */
+  public getAndroidDistributionFile(
+    distributionDir: string
+  ): string | undefined {
+    const files = fs.readdirSync(distributionDir);
```

Comment:
Can this be part of the target config? Directory walking operations are quite expensive and you seem to be doing a lot of them.